### ISSUE TITLE
[BUG] Batch step can timeout when there are a lot of messages to process

### ISSUE THREAD

**Issue by arnejduranovic** [CDCgov/prime-reportstream#10305]
## Describe the bug
SimpleReport sent hundreds of thousands of messages to RS at the same time several times last week and when the batch step was executed, it failed to process everything. It kept failing (timing out) until all the messages were eventually delivered. See below for more details:

We got 147,174 covid results from Simple Report's CSV uploader on June 30th, 2023.   All the results come from the same provider in NY with patients in NJ and have tests that were taken since last year. 

![image](https://github.com/CDCgov/prime-reportstream/assets/118766341/f1357895-ad7f-47c2-906a-9b10c6a6f179)

## Impact on ReportStream
This bug causes Pager duty to be very loud, but the messages eventually get delivered.

## Steps to reproduce
I would not recommend reproducing this error, but if you wanted to, just send thousands of messages at the same time to RS.

## Expected behavior
The Batch function should not time out. It should probably limit itself to a specific amount of messages to ensure it never times out.

## Logs
See azure function logs for June 30th on production instance.

**Comment by JessicaWNava** [comment]
So it looks like there are two places that are downloading the blob for UP, one in BatchFunction.batchUniversalData and once further down the line in WorkflowEngine.createHeader called from WorkflowEngine.handleBatchEvent. Currently we do this one by one as Kalish was saying in standup. This appears to be eating up the most time, so one optimization we could do is trying to run this in parallel. Besides that, it turns out that the BatchDecider actually does make it so that all the messages are processed and sent no matter what in one go. So we can lower the number of messages per batch without consequences. Thoughts on how to proceed @arnejduranovic ?
@avnieldravid please disregard what I said in standup this morning, it was incorrect.

**Comment by arnejduranovic** [comment]
Since we cannot be sure parallelization will fix this "bug", My vote is to create a new ticket to implement said parallelization and point this one to the new ticket. 

Prior to working on that ticket, we should also create and work on a new ticket for setting up a functional test case that tries to reproduce this bug - so that when we implement the parallel download we can verify it fixes the issue. 

Thoughts @mkalish ?

**Comment by mkalish** [comment]
@arnejduranovic @JessicaWNava 
I think trying to get a functional test would be good though it might be difficult as a quick search didn't reveal a way to change the timeout locally, so we would be stuck with the 10 minutes.  We could also look into creating some better logging around the download process and create some azure charts to show the before and after.  I'm thinking something that logs the total time spent downloading per batch execution.

Definitely agree that this is not really a bug and needs to be split into it's own work

**Comment by arnejduranovic** [comment]
So my vote is to then create the following tickets:
- Functional test ticket that will investigate an appropriate way to test, if we are stuck with a 10min test then perhaps its one we trigger manually or somehow outside of CI?
- Download logging and azure charts
- Parallelization 

I also think we should make an epic and assign all three stories to it, I can see more work coming out of this potentially

**Comment by JessicaWNava** [comment]
Created the tickets:
https://app.zenhub.com/workspaces/platform-6182b02547c1130010f459db/issues/gh/cdcgov/prime-reportstream/10691
https://app.zenhub.com/workspaces/platform-6182b02547c1130010f459db/issues/gh/cdcgov/prime-reportstream/10692
https://app.zenhub.com/workspaces/platform-6182b02547c1130010f459db/issues/gh/cdcgov/prime-reportstream/10693
